test/analysis/typeComplexityAnalysis.test.ts

```diff
@@ -0,0 +1,328 @@
+            query = `Query { human(id: 1) { id, name, appearsIn } }`;
+            expect(getQueryTypeComplxity(query, typeWeights)).toBe(3); // Query 1 + human/character 1 + appearsIn/episode
+            // argument passed in as a variable
+            query = `Query { hero(episode: $ep) { id, name } }`;
+            expect(getQueryTypeComplxity(query, typeWeights)).toBe(2); // Query 1 + hero/character 1
+        });
+
+        test('with fragments', () => {
+            query = `
+            Query {
+                leftComparison: hero(episode: EMPIRE) {
+                  ...comparisonFields
+                }
+                rightComparison: hero(episode: JEDI) {
+                  ...comparisonFields
+                }
+              }
+              
+              fragment comparisonFields on Character {
+                name
+                appearsIn
+              }
+            }`;
+            expect(getQueryTypeComplxity(query, typeWeights)).toBe(5); // Query 1 + 2*(character 1 + appearsIn/episode 1)
+        });
+
+        test('with inline fragments', () => {
+            query = `
+            Query {
+                hero(episode: EMPIRE) {
+                    name
+                    ... on Droid {
+                        primaryFunction
+                    }
+                    ... on Human {
+                        homeplanet
+                    }
+                }
+            }`;
+            expect(getQueryTypeComplxity(query, typeWeights)).toBe(2); // Query 1 + hero/character 1)
+        });
+
+        /**
+         *
+         * With type complexity analysis, all objects returned count towards the total complexity.
+         * For example, the cost of querying for 5 friends is 5. I do not have any clue how we would know
+         * to look for the argument 'first' to know, before running the query, how many objects are expected to be returned.
+         *
+         * Anouther example, if we queried the 'Search' type with some string argument, the returned number of objects
+         * could be very large. Our algorithm will need to know what limit is set for the returned data (limit 100 search results
+         * for example) and then account for that response to caculate the complexity. That information is in the resolvers. We
+         * have no access to the resolvers.
+         *
+         * Some user configuration will be needed unless someone has bright ideas.
+         */
+        // ? type weigts are variable, not sure how to calculate this.
+        test('with lists', () => {
+            query = `
+            Query { 
+                human(id: 1) { 
+                    name, 
+                    friends { 
+                        name 
+                    } 
+                }
+            }`;
+            expect(getQueryTypeComplxity(query, typeWeights)).toBe(7); // 1 Query + 1 human/character +  5 friends/character
```

Comment:
We should add a FIXME to address this once we determine how to calculate the cost of list lengths.